credoai/evaluators/shap.py

```diff
@@ -0,0 +1,141 @@
+
+    Parameters
+    ----------
+    samples_ind : Optional[List[int]], optional
+        List of row numbers representing the samples for which to extract individual
+        shapley values , by default None
+    """
+
+    def __init__(self, samples_ind: Optional[List[int]] = None):
+
+        super().__init__()
+        self.samples_ind = samples_ind
+
+    name = "Shap"
+    required_artifacts = ["assessment_data", "model"]
+
+    def _setup(self):
+        self.X = self.assessment_data.X
+        self.model = self.model
+        return self
+
+    def _validate_arguments(self):
+        check_requirements_existence(self)
+
+    def evaluate(self):
+        ## Overall stats
+        self._setup_shap()
+        res = self._get_overall_shap_contributions()
+        self.results = [TableContainer(res, **self.get_container_info())]
+
+        ## Sample specific results
+        if self.samples_ind:
+            labels = {"ordered_feature_names": self.shap_values.feature_names}
+            ind_res = self._get_mult_sample_shapley_values()
+            self.results += [
+                TableContainer(ind_res, **self.get_container_info(labels=labels))
+            ]
+        return self
+
+    def _setup_shap(self):
+        """
+        Setup the explainer given the model and the feature dataset
+        """
+        explainer = Explainer(self.model.predict, self.X)
+        self.shap_values = explainer(self.X)
+        return self
+
+    def _get_overall_shap_contributions(self) -> DataFrame:
+        """
+        Calculate overall SHAP contributions for a dataset.
+
+        The output of SHAP package provides Shapley values for each sample in a
+        dataset. To summarise the contribution of each feature in a dataset, the
+        samples contributions need to be aggregated.
+
+        For each of the features, this method provides: mean, max, minimum and the
+        mean of the absolute value of the samples shapley values.
+
+        Returns
+        -------
+        DataFrame
+            Summary of the shapley values across the full dataset.
+        """
+        values_df = DataFrame(self.shap_values.values)
+        values_df.columns = self.shap_values.feature_names
+        stats_1 = values_df.apply(["mean"]).T
+        stats_2 = values_df.apply(lambda x: mean(abs(x)))
+        stats_2.name = "mean(|x|)"
+        final = concat([stats_1, stats_2], axis=1)
+        final = final.sort_values("mean(|x|)", ascending=False)
+        final.name = "Summary of Shap statistics"
+
+        return final
+
+    def _get_mult_sample_shapley_values(self) -> DataFrame:
+        """
+        Return shapley values for multiple samples from the dataset.
+
+        Returns
+        -------
+        DataFrame
+            Columns:
+                values -> shap values
+                ref_value -> Reference value for the shap values
+                    (generally the same across the dataset)
```

Comment:
What exactly is this? The mean shap value for the given feature or expected value for a sample (i.e. mean prediction output for all samples)? 

Can we clarify the comment by pulling from SHAP's documentation?

Reply:
This are the actual shap values. As in for 2 samples and 3 features

sample1 -> shap_val_1, ..., shap_val_3
sample2 -> shap_val_1, ..., shap_val_3

Reply:
I was referring to the reference value. That's something related to the background data set?

Reply:
Yeah as per our discussion yesterday, I can't find any specific reference in the docs as to how it's calculated.